### PR TITLE
Fix BFloat16 logits returned as garbage in Logits::Get()

### LINKED ISSUE #2202: BFloat16 logits returned as garbage — Logits::Get() only casts Float16 to float32
### Describe the bug

For models with **BFloat16** logits output, `onnxruntime-genai` returns garbage logits, producing a wrong/incoherent first token and degenerate generation. The identical model exported in **Float16** works correctly.

### Root cause

`src/models/logits.cpp` (`Logits::Get()`) only converts the model's raw logits to `float32` when the output type is **Float16**. For **BFloat16** it skips the conversion, then `WrapTensor<float>(...)` reinterprets the raw 2-byte bf16 bytes as 4-byte float32, corrupting every logit.

Two places are gated on `Float16` only:

```cpp
// ~line 47 — fp32 staging buffer is allocated only for Float16
if (type_ == Ort::TypeToTensorType<Ort::Float16_t>)
  logits_of_last_token_fp32_ = OrtValue::CreateTensor<float>(...);

// ~line 73 — Cast to float32 runs only for Float16
if (type_ == Ort::TypeToTensorType<Ort::Float16_t>) {
  Cast(*logits_of_last_token, logits_of_last_token_fp32_, *model_.p_device_inputs_, Ort::TypeToTensorType<float>);
  logits_of_last_token = logits_of_last_token_fp32_.get();
}
```

For a bf16 model the `Cast` never runs, so `logits_` ends up wrapping raw bf16 bytes as `float`.

### How I verified

Isolated a bf16 vs fp16 model (Gemma-style decoder, vocab 262144):
- HF bf16 reference, raw ORT `sess.run` of the bf16 decoder, and genai's own dumped model-output logits **all agree** (argmax 496, top-5 `[496,886,506,9079,992]`).
- genai's `get_logits()` for the **same** bf16 model returns argmax 4539 — a value that does **not** correspond to any position in the model's actual output. The corruption is purely in the logits read-out, not the model compute or the ORT kernels.

### Proposed fix

Treat `BFloat16` the same as `Float16` in both spots, and add a bf16→f32 conversion path. The existing `Cast()` helper already falls back to a CPU bf16→f32 cast; I also added a CUDA `LaunchBf16ToFp32` kernel so the cast stays on-device.

```diff
--- a/src/models/logits.cpp
+++ b/src/models/logits.cpp
-    if (type_ == Ort::TypeToTensorType<Ort::Float16_t>)
+    if (type_ == Ort::TypeToTensorType<Ort::Float16_t> || type_ == Ort::TypeToTensorType<Ort::BFloat16_t>)
       logits_of_last_token_fp32_ = OrtValue::CreateTensor<float>(...);
...
-  // Convert from float16 to float32 if necessary
-  if (type_ == Ort::TypeToTensorType<Ort::Float16_t>) {
+  // Convert from float16/bfloat16 to float32 if necessary
+  if (type_ == Ort::TypeToTensorType<Ort::Float16_t> || type_ == Ort::TypeToTensorType<Ort::BFloat16_t>) {
     Cast(*logits_of_last_token, logits_of_last_token_fp32_, *model_.p_device_inputs_, Ort::TypeToTensorType<float>);
     logits_of_last_token = logits_of_last_token_fp32_.get();
   }
```

Plus a CUDA bf16→f32 case (`src/cuda/interface.cpp` `Cast`, `src/cuda/model_kernels.cu` / `kernels.h` `LaunchBf16ToFp32`).

After the fix, the bf16 model's first-token argmax is 496 (matches HF/fp16) and generation is byte-identical to fp16.

### Urgency

Medium — blocks deploying any bf16 model via genai. fp16 is an available workaround but loses bf16's wider dynamic range.

### Platform / Version

- onnxruntime-genai built from source, CUDA EP, Linux.
- Reproduces independent of EP/kernel dispatch (model compute is correct; only the logits read-out is affected).

I have a tested fix and can open a PR if useful.

### PR DESCRIPTION
**Pull request by justinchuby**
### Summary

`Logits::Get()` only converted the model's raw logits to `float32` when the output type was **Float16**. For **BFloat16** the conversion was skipped, and the subsequent `WrapTensor<float>` reinterpreted the raw 2-byte bf16 values as 4-byte float32 — corrupting every logit (wrong argmax, incoherent generation). The identical model in Float16 worked correctly.

Fixes #2202.

### Changes

- `src/models/logits.cpp`: treat `BFloat16` the same as `Float16` in both the fp32 staging-buffer allocation and the `Cast` to float32.
- `src/cuda/model_kernels.cu` + `src/cuda/kernels.h`: add `LaunchBf16ToFp32` (on-device bf16→f32 conversion).
- `src/cuda/interface.cpp`: route `BFloat16 → float` through the new kernel in the CUDA `Cast`. (The CPU `Cast` already supported bf16→f32, so the generic `Cast` helper otherwise falls back to a host round-trip.)

### Verification

On a bf16 decoder (vocab 262144), the first-token argmax now matches the Float16 / HuggingFace reference, and greedy generation is byte-identical to the fp16 export. Before the fix, genai returned argmax 4539 (a value that did not correspond to any position in the model's actual output); after, it returns 496 (correct).

### CODE REVIEW

**Review by kunal-vaishnavi — DISMISSED** [review]
Typically, we have upcasted the logits to float32 inside the model itself.

https://github.com/microsoft/onnxruntime-genai/blob/13adddebb149afa9dd344c165483d567e7fb4a2b/src/python/py/models/builders/base.py#L385-L387

But I think it is still good to have the upcasting option inside ORT GenAI as a fallback.

File: src/cuda/model_kernels.cu [hunk 1]
```diff
@@ -128,6 +129,19 @@ void LaunchFp32ToFp16(const float* fp32, uint16_t* fp16, int count, cudaStream_t
...
 }
 
+__global__ void ConvertBf16ToFp32(const __nv_bfloat16* src, float* dst, int count) {
+  int idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx < count)
+    dst[idx] = __bfloat162float(src[idx]);
+}
+
+void LaunchBf16ToFp32(const uint16_t* bf16, float* fp32, int count, cudaStream_t stream) {
+  int block_size = 256;
```
Comment (baijumeswani) [inline src/cuda/model_kernels.cu]:
```suggestion
  constexpr int block_size = 256;
```

File: src/cuda/model_kernels.cu [hunk 2]
```diff
@@ -128,6 +129,19 @@ void LaunchFp32ToFp16(const float* fp32, uint16_t* fp16, int count, cudaStream_t
...
+  int num_blocks = (count + block_size - 1) / block_size;
```
Comment (baijumeswani) [inline src/cuda/model_kernels.cu]:
```suggestion
  const int num_blocks = (count + block_size - 1) / block_size;
```

File: src/models/logits.cpp [hunk 3]
```diff
@@ -44,7 +44,7 @@ DeviceSpan<float> Logits::Get() {
     // create new OrtValue for logits_of_last_token and use output_last_tokens_ to hold it
     output_last_tokens_ = OrtValue::CreateTensor(model_.p_device_inputs_->GetAllocator(), shape_last, type_);
 
-    if (type_ == Ort::TypeToTensorType<Ort::Float16_t>)
+    if (type_ == Ort::TypeToTensorType<Ort::Float16_t> || type_ == Ort::TypeToTensorType<Ort::BFloat16_t>)
       logits_of_last_token_fp32_ = OrtValue::CreateTensor<float>(model_.p_device_inputs_->GetAllocator(), shape_);
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/models/logits.cpp:48]: @baijumeswani does this apply?

[the remark this answers is not in the document]
Comment (baijumeswani) [reply src/models/logits.cpp:48]: Will need to investigate. This is not directly related to your PR so I think you can avoid for now. Will investigate this separately.

**Comment by justinchuby** [comment]
@kunal-vaishnavi @baijumeswani could you approve? Thanks.

**Comment by baijumeswani** [comment]
@justinchuby can you merge main to trigger the integration test pipeline?

**Review by kunal-vaishnavi — APPROVED** [review]

**Comment by kunal-vaishnavi** [comment]
We need to sync with the main branch of ORT GenAI to get the changes needed to run the integration tests.

**Comment by justinchuby** [comment]
Doesn't seem to be triggered 🤔

**Review by natke — APPROVED** [review]
